### PR TITLE
Support capture screen

### PR DESCRIPTION
**Pull request by YoungeunKim**
:Release Notes:
Support capture in ares-device

:Detailed Notes:
Implement capture screen feature to developer watch target display
- Add ares-device --capture-screen(-c)
- Add capture option description in ares-device help  
- Update positive/negative TC for capture option 

:Testing Performed:
1. Pass unit test on ose and auto
2. Pass eslint
3. Verify with CLI commands

- Check capture file generated with _targetname__display0__time_.png
$ ./ares-device.js -c 
Create rpi4_display0_20210115095912.png to /home/ye0607kim/project/github/ares-cli/bin
Success

- Give display option and capture file name should be _targetname__display1__time_.png
$./ares-device.js -c -dp 1
Create rpi4_display1_20210115095826.png to /home/ye0607kim/project/github/ares-cli/bin
Success
                                      
- Give directory path with display option
$ ./ares-device.js -c webOSCap -dp 1
Create rpi4_display1_20210115095944.png to /home/ye0607kim/project/github/ares-cli/bin/webOSCap
Success

- Give file name and format(png, bmp, jpg) and check the generated file name & format
$ ./ares-device.js -c webOSCap/test1.bmp _[or test1.jpg or test1.png]_
Create test1.bmp  _[or test1.jpg or test1.png]_ to /home/ye0607kim/project/github/ares-cli/bin/webOSCap
Success

- Check /tmp/aresCapture has been removed from device target after command working completed. 

:Issues Addressed:
[PLAT-125564] Implement ares-device --capture-screen

### CODE REVIEW

File: lib/device.js [hunk 1]
```diff
@@ -251,6 +256,177 @@ const util = require('util'),
...
+                                return setImmediate(next, errHndl.changeErrMsg("FAILED_CALL_LUNA", lineObj.errorText), {});
+                            }
+                        }, next);
+                    }
+                });
+            }
+
+            function _copyToLocal(next) {
+                log.info("device#capture#_copyToLocal()");
+
```
Comment (YoungeunKim) [inline lib/device.js:404]: Remove new line

**Comment by YoungeunKim** [comment]
Code 👍 
Verified 👍

File: files/help/ares-device.help [hunk 2]
```diff
@@ -16,9 +16,15 @@
     },
     "options" : {
         "cmdOpt":"option",
-        "default" : ["system", "session", "device", "device-list", "help", "version", "verbose"],
+        "default" : ["system", "session", "capture", "display", "device", "device-list", "help", "version", "verbose"],
         "system" : "-i, --system-info @TAB@ Display the device system information",
         "session" : "-s, --session-info @TAB@ Display the device session information",
+        "capture" : [
+            "-c, --capture <OUTPUT_PATH> @TAB@ Capture screen and save the file to local",
```
Comment (ssuminahn) [inline files/help/ares-device.help:23]: Capture screen and save captured file in host PC.
I think this is proper..

File: bin/ares-device.js [hunk 3]
```diff
@@ -78,6 +81,8 @@ if (argv.level) {
 
 const options = {
     device: argv.device,
+    display : argv.display || 0,
+    inputPath : argv.argv.remain[0] || null,
```
Comment (ssuminahn) [inline bin/ares-device.js:85]: In help message, this is named output path.
I think the two are the same. So I want the two to have the same name.

File: lib/base/novacom.js [hunk 4]
```diff
@@ -72,7 +72,7 @@ const fs = require('fs'),
                 if (err) {
                     setImmediate(next, err);
                 } else {
-                    log.info("Resolver#load()", "devices:", resolver.devices);
+                    log.verbose("Resolver#load()", "devices:", resolver.devices);
```
Comment (ssuminahn) [inline lib/base/novacom.js:75]: Do you want to print the log in verbose?

File: lib/device.js [hunk 5]
```diff
@@ -7,7 +7,12 @@
 const util = require('util'),
     async = require('async'),
     npmlog = require('npmlog'),
+    fs = require('fs'),
+    path = require('path'),
+    mkdirp = require('mkdirp'),
+    chalk = require('chalk'),
     luna = require('./base/luna'),
+    pullLib = require('./pull'),
```
Comment (ssuminahn) [inline lib/device.js:15]: I think it would be better if this is the last one.

File: lib/device.js [hunk 6]
```diff
@@ -251,6 +256,174 @@ const util = require('util'),
...
+                            addr = target.lunaAddr.captureCompositorOutput,
+                            param = {
+                                // luna param
+                                subscribe: false,
+                                output : options.sourcePath,
+                                format: options.capFormat,
+                                displayId : options.display
+                            };
+
+                        luna.sendWithoutErrorHandle(options, addr, param, function(lineObj, next) {
```
Comment (ssuminahn) [inline lib/device.js:387]: why do you use sendWithoutErrorHandle?
if some error occurs, we don't check them.

File: lib/device.js [hunk 7]
```diff
@@ -251,6 +256,174 @@ const util = require('util'),
...
+
+                function __pad(number, length) {
+                    let str = '' + number;
+                    while (str.length < length) {
+                        str = '0' + str;
+                    }
+                    return str;
+                }
+
+                const fileName = options.session.target.name + "_" +"display" + options.display+ "_" + __getDate() +".png";
```
Comment (ssuminahn) [inline lib/device.js:365]: "_" +"display"

"_display" 
or 
"_" + "dispplay" (missing a space)

File: lib/device.js [hunk 8]
```diff
@@ -251,6 +256,174 @@ const util = require('util'),
...
+            function _makeSession(next) {
+                options.nReplies = 1; // -n 1
+                options.session = new novacom.Session(options.device, next);
+            }
+
+            function _makeCaptureOption(next) {
+                log.info("device#capture#_makeCaptureOption()");
+                let capFilename = "";
+                let capFormat = "PNG"; // PNG is default format
+                let outputDirPath;
```
Comment (ssuminahn) [inline lib/device.js:300]: let capFilename = "",
capFormat = "PNG",
outputDirPath;

is better...I think.

File: lib/device.js [hunk 9]
```diff
@@ -251,6 +256,174 @@ const util = require('util'),
...
+                    const parseBase = path.parse(options.inputPath).base;
+                    let parseExt = path.parse(options.inputPath).ext;
+                    parseExt = parseExt.split('.').pop();
+
+                    log.info("device#capture#_makeCaptureOption():" + " dir name : " + parseDirPath 
+                        + " ,file name : " + parseBase +" ,inputFormat : " +  parseExt);
+
+                    // if given path has extension
+                    if (parseBase && parseExt) {
+                        if ((parseExt === "png") || (parseExt === "bmp") || (parseExt === "jpg") ){
```
Comment (ssuminahn) [inline lib/device.js:317]: is () needed?

File: lib/device.js [hunk 10]
```diff
@@ -251,6 +256,174 @@ const util = require('util'),
...
+                        }
+                    });
+                } else {
+                    pullLib.pull(options, next);
+                }
+            }
+
+            function _removeCaptureDir(next) {
+                log.info("device#capture#_removeCaptureDir()");
+                const cmd = '/bin/rm -rf ' + options.capDirPath;
```
Comment (ssuminahn) [inline lib/device.js:418]: is /bin/ needed?

File: lib/pull.js [hunk 11]
```diff
@@ -69,7 +69,9 @@ const fs = require('fs'),
                     let cmd = "[ -f " + sourcePath + " ] && echo 'f' || echo 'nf'";
 
                     session.run(cmd, null, writeIsFile, null, function(){
-                        console.log("Copying data ....");
+                        if(!options.silent) {
```
Comment (ssuminahn) [inline lib/pull.js:72]: Does this not affect the existing case?
Already relate cases are in qwa, qsdk test case.

**Review by ssuminahn — CHANGES_REQUESTED** [review]
verify :+1: 

code is needed discussion with us.

- I think you would check help message one more and refer other command's help messaage. 
- I don't want to shorten the variable name. For example, capFileName should be change to captureFileName. 

I haven't reviewed through all of the logic. First of all, I'll comment like this, take a closer look at the logic, and add more comments.

File: files/help/ares-device.help [hunk 12]
```diff
@@ -43,6 +49,12 @@
             "",
             "# Display session information",
             "ares-device -s -d DEVICE",
+            "",
+            "# Display session information",
```
[the remark this answers is not in the document]
Comment (YoungeunKim) [reply files/help/ares-device.help:53]: Done.

File: files/help/ares-device.help [hunk 13]
```diff
@@ -43,6 +49,12 @@
...
+            "ares-device -s -d DEVICE",
+            "",
+            "# Capture screen as screen.png of display 1",
```
[the remark this answers is not in the document]
Comment (YoungeunKim) [reply files/help/ares-device.help:56]: Done.

File: files/help/ares-device.help [hunk 14]
```diff
@@ -16,9 +16,15 @@
     },
     "options" : {
         "cmdOpt":"option",
-        "default" : ["system", "session", "device", "device-list", "help", "version", "verbose"],
+        "default" : ["system", "session", "capture", "display", "device", "device-list", "help", "version", "verbose"],
         "system" : "-i, --system-info @TAB@ Display the device system information",
         "session" : "-s, --session-info @TAB@ Display the device session information",
+        "capture" : [
+            "-c, --capture <OUTPUT_PATH> @TAB@ Capture screen and save the file to local",
+            "@TAB@ Use OUTPUT_PATH as the output directory or file path",
```
[the remark this answers is not in the document]
Comment (YoungeunKim) [reply files/help/ares-device.help:24]: Good idea :+1: 
I updated it.

File: files/help/ares-device.help (hunk 2 above)
Comment (YoungeunKim) [reply files/help/ares-device.help:23]: I replaced it host machine, which is used in ares-pull help.
Done.

File: bin/ares-device.js (hunk 3 above)
Comment (YoungeunKim) [reply bin/ares-device.js:85]: I agree with you. Fixed. Done.

File: files/help/ares-device.help (hunk 2 above)
[the remark this answers is not in the document]
Comment (YoungeunKim) [reply files/help/ares-device.help:23]: I agree with you. I thought like you, but we already determine the option in design review, so I didn't suggest the change during implementing the code. 

Why don't you -c (--capture-screen) ?

File: lib/base/novacom.js (hunk 4 above)
Comment (YoungeunKim) [reply lib/base/novacom.js:75]: Yes, If you want to keep it to "info" level, I'll change it to "info" as original.

File: lib/device.js (hunk 5 above)
Comment (YoungeunKim) [reply lib/device.js:15]: Fixed. Done.

File: lib/device.js (hunk 6 above)
Comment (YoungeunKim) [reply lib/device.js:387]: We can check "luna-send error handle" in luna.sendWithoutErrorHandle, too.
Please verify 
{code}
$ ./ares-device.js -c -dp 10
ares-device ERR! CLI: luna-send command failed <ERR_INVALID_DISPLAY> 
{code}

I understand that the luna.send is better than sendWithoutErrorHandle. Right?

File: lib/device.js (hunk 7 above)
Comment (YoungeunKim) [reply lib/device.js:365]: I added space. Done.

File: lib/device.js (hunk 9 above)
Comment (YoungeunKim) [reply lib/device.js:317]: Yes, is needed.
If user gives path /webOS/test need to check that "test" path is directory path or file path.

File: lib/device.js (hunk 10 above)
Comment (YoungeunKim) [reply lib/device.js:418]: Yes. usually in CLI, make command, which is use for target device, add /bin/. 
{code}
options.session.run('/bin/rm -f "' + devicePkgPath + '"', null, null, null, next); 
{code}

File: lib/device.js (hunk 8 above)
Comment (YoungeunKim) [reply lib/device.js:300]: Done.

File: lib/pull.js (hunk 11 above)
Comment (YoungeunKim) [reply lib/pull.js:72]: Not affect it. I've already check it. 
`$ ./ares-pull.js /tmp/testPull.txt .
Copying data ....
Pull: /tmp/testPull.txt -> ./testPull.txt
1 file(s) pulled
138 KB/s (4096 bytes in 0.029s)
ye0607kim@ye0607kim-linux:~/project/github/ares-cli/bin$ ./ares-pull.js /tmp/testPull.txt . -i
Copying data ....
1 file(s) pulled
103 KB/s (4096 bytes in 0.039s)`

**Comment by YoungeunKim** [comment]
Code :+1: 
Verified :+1: 

@ssuminahn , Let's have a discussion about "unresolved conversation" s.

File: lib/device.js [hunk 15]
```diff
@@ -362,36 +360,33 @@ const util = require('util'),
                     return str;
                 }
 
-                const fileName = options.session.target.name + "_" +"display" + options.display+ "_" + __getDate() +".png";
+                const fileName = options.session.target.name + "_" + "display" + options.display+ "_" + __getDate() + ".png";
```
Comment (YoungeunKim) [inline lib/device.js:363]: Add space options.display +

File: files/help/ares-device.help (hunk 2 above)
[the remark this answers is not in the document]
Comment (ssuminahn) [reply files/help/ares-device.help]: I agree -c (--capture-screen)

File: lib/device.js (hunk 15 above)
Comment (YoungeunKim) [reply lib/device.js]: Done.

File: files/help/ares-device.help (hunk 2 above)
[the remark this answers is not in the document]
Comment (YoungeunKim) [reply files/help/ares-device.help]: I change it to --capture-screen.

**Review by YoungeunKim — COMMENTED** [review]
Code :+1: 
Verified :+1:

File: lib/device.js [hunk 16]
```diff
@@ -251,6 +256,169 @@ const util = require('util'),
...
+                _makeSession,
+                _makeCaptureOption,
+                _captureFile,
+                _copyToLocal
+            ],  function(err, results) {
+                log.verbose("device#capture()", "err: ", err, "results:", results);
+                const resultTxt = "Create " + chalk.green(options.captureFileName) + " to " + options.destinationPath +"\nSuccess";
+
+                // clean up /tmp/capture directory in target device
+                if (options.targetCaptureDir) {
```
Comment (ssuminahn) [inline lib/device.js:280]: Doesn't check if the actual file exists, is it correct to delete it unconditionally if it is in option?

File: lib/device.js [hunk 17]
```diff
@@ -251,6 +256,169 @@ const util = require('util'),
...
+                    } else if (parseBase && (parseExt === "")) {
+                        // it's directory path. Use it as destination dir Path
+                        captureFileName = _createDateFileName();
+                        destinationPath = path.resolve(options.outputPath);
+                    }
+                }
+
+                options.captureFormat = captureFormat;
+                options.captureFileName = captureFileName;
+                options.captureDirPath = "/tmp/capture",
```
Comment (ssuminahn) [inline lib/device.js:333]: How about writing the folder name in more detail in case it won't be deleted?

File: lib/device.js [hunk 18]
```diff
@@ -251,6 +256,169 @@ const util = require('util'),
...
+                _makeCaptureOption,
+                _captureFile,
+                _copyToLocal
+            ],  function(err, results) {
+                log.verbose("device#capture()", "err: ", err, "results:", results);
+                const resultTxt = "Create " + chalk.green(options.captureFileName) + " to " + options.destinationPath +"\nSuccess";
+
+                // clean up /tmp/capture directory in target device
+                if (options.targetCaptureDir) {
+                    _removeCaptureDir(function finish (removeErr){
```
Comment (ssuminahn) [inline lib/device.js:281]: function finish(removeErr) {

File: lib/device.js [hunk 19]
```diff
@@ -251,6 +256,169 @@ const util = require('util'),
...
+                if (options.outputPath === null){
+                    captureFileName = _createDateFileName();
+                    destinationPath = path.resolve('.');
+                } else {
+                    // Get directory path, file path, file Extenstion from given path
+                    const parseDirPath = path.dirname(options.outputPath),
+                        parseBase = path.parse(options.outputPath).base;
+
+                    let parseExt = path.parse(options.outputPath).ext;
+                    parseExt = parseExt.split('.').pop();
```
Comment (ssuminahn) [inline lib/device.js:309]: Please check this line.
Is this new assign paresExt?

File: lib/device.js [hunk 20]
```diff
@@ -251,6 +256,169 @@ const util = require('util'),
...
+
+                    log.info("device#capture#_makeCaptureOption():" + " dir name : " + parseDirPath 
+                        + " ,file name : " + parseBase +" ,inputFormat : " +  parseExt);
```
Comment (ssuminahn) [inline lib/device.js:312]: need a space

File: lib/device.js (hunk 19 above)
Comment (ssuminahn) [inline lib/device.js:309]: How about use path.extname()?
https://nodejs.org/api/path.html#path_path_extname_path

File: lib/device.js [hunk 21]
```diff
@@ -251,6 +256,169 @@ const util = require('util'),
...
+                    const dateFormat = __pad(curDate.getFullYear(), 2)
+                        +  __pad((curDate.getMonth()+1), 2)
+                        + __pad(curDate.getDate(), 2)
+                        + __pad(curDate.getHours(), 2)
+                        + __pad(curDate.getMinutes(), 2)
+                        + __pad(curDate.getSeconds(), 2);
+                    return dateFormat;
+                }
+
+                function __pad(number, length) {
```
Comment (ssuminahn) [inline lib/device.js:355]: what is pad mean?

File: lib/device.js [hunk 22]
```diff
@@ -251,6 +256,169 @@ const util = require('util'),
...
+                        str = '0' + str;
+                    }
+                    return str;
+                }
+
+                const fileName = options.session.target.name + "_" + "display" + options.display+ "_" + __getDate() + ".png";
+                return fileName;
+            }
+
+            function _captureFile(next) {
```
Comment (ssuminahn) [inline lib/device.js:367]: how about captureScreen?

File: lib/device.js [hunk 23]
```diff
@@ -251,6 +256,169 @@ const util = require('util'),
...
+                                // luna param
+                                subscribe: false,
+                                output : options.sourcePath,
+                                format: options.captureFormat,
+                                displayId : options.display
+                            };
+
+                        luna.send(options, addr, param, function(lineObj, next) {
+                            log.silly("device#capture#_captureFile():", "lineObj:", lineObj);
+                            if (lineObj.returnValue) {
```
Comment (ssuminahn) [inline lib/device.js:387]: we need a error handle when lineObj.returnValue is null or fail return or occured error.

File: lib/device.js [hunk 24]
```diff
@@ -251,6 +256,169 @@ const util = require('util'),
...
+                        } else {
+                            pullLib.pull(options, next);
+                        }
+                    });
+                } else {
+                    pullLib.pull(options, next);
+                }
+            }
+
+            function _removeCaptureDir(next) {
```
Comment (ssuminahn) [inline lib/device.js:411]: removeTmpDir

File: lib/device.js [hunk 25]
```diff
@@ -251,6 +256,169 @@ const util = require('util'),
...
+                            if (lineObj.returnValue) {
+                                log.info("device#capture#_captureFile():","Capture file in target : " + lineObj.output);
+                                next(null, {});
+                            }
+                        }, next);
+                    }
+                });
+            }
+
+            function _copyToLocal(next) {
```
Comment (ssuminahn) [inline lib/device.js:396]: copyFileToLocal

**Review by ssuminahn — CHANGES_REQUESTED** [review]
@YoungeunKim 
I have posted comments.
After you check them, let's discuss the parts that need to be discussed separately.

File: files/help/ares-device.help [hunk 26]
```diff
@@ -11,14 +11,20 @@
     },
     "description" : {
         "default": [
-            "This command displays the device information."
+            "This command displays the device information.",
```
[the remark this answers is not in the document]
Comment (YoungeunKim) [reply files/help/ares-device.help:14]: I considered to update the description, but it's hard to make general(common) sentence.
"This command displays the device's system/session information and capture file".
-> I think that this sentence is not proper, I think. 

If you suggest description, I'll update.

File: spec/jsSpecs/ares-device.spec.js [hunk 27]
```diff
@@ -86,3 +90,122 @@ describe(aresCmd, function() {
...
+    it('Capture', function(done) {
+        exec(cmd + ` --capture`, function (error, stdout) {
+            const curDate = new Date();
+            expect(stdout).toContain(options.device);
+            expect(stdout).toContain(curDate.getFullYear());
+            expect(stdout).toContain("display0");
+            expect(stdout).toContain(".png");
+            expect(stdout).toContain(path.resolve('.'));
+
+            const reg = new RegExp("[A-Za-z0-9]*_[dispaly0-9]*_[0-9]*.png");
```
[the remark this answers is not in the document]
Comment (YoungeunKim) [reply spec/jsSpecs/ares-device.spec.js:113]: [A-Za-z0-9]*_display[0-9]_[0-9]*.png is correct. I changed it. 
Thanks.

File: spec/jsSpecs/ares-device.spec.js [hunk 28]
```diff
@@ -86,3 +90,122 @@ describe(aresCmd, function() {
...
+    afterEach(function(done) {
+        common.removeOutDir(captureDirPath);
+        done();
+    });
+
+    it('Capture', function(done) {
+        exec(cmd + ` --capture`, function (error, stdout) {
+            const curDate = new Date();
+            expect(stdout).toContain(options.device);
+            expect(stdout).toContain(curDate.getFullYear());
```
[the remark this answers is not in the document]
Comment (YoungeunKim) [reply spec/jsSpecs/ares-device.spec.js:108]: Date.getFullYear() is invalid.
:  Uncaught exception: TypeError: Date.getFullYear is not a function

File: spec/jsSpecs/ares-device.spec.js [hunk 29]
```diff
@@ -86,3 +90,122 @@ describe(aresCmd, function() {
...
+            expect(stdout).toContain(path.resolve('.'));
+
+            const reg = new RegExp("[A-Za-z0-9]*_[dispaly0-9]*_[0-9]*.png");
+            const arr = stdout.match(reg);
+            const filePath = path.join(path.resolve('.'), arr[0]);
+            console.log("capture file name : " + arr[0]);
+            expect(fs.existsSync(filePath)).toBe(true);
+
+            // remove created capture file
+            common.removeOutDir(filePath);
```
[the remark this answers is not in the document]
Comment (YoungeunKim) [reply spec/jsSpecs/ares-device.spec.js:120]: The file name is generated depends on date/time. If I remove it on afterEash, 
I'll keep this value in globally. 

So, I hope remove the file in exec() command.

File: spec/jsSpecs/ares-device.spec.js [hunk 30]
```diff
@@ -86,3 +90,122 @@ describe(aresCmd, function() {
...
+            expect(stdout).toContain(options.device);
+            expect(stdout).toContain(curDate.getFullYear());
+            expect(stdout).toContain("display0");
+            expect(stdout).toContain(".png");
+            expect(stdout).toContain(path.resolve('.'));
+
+            const reg = new RegExp("[A-Za-z0-9]*_[dispaly0-9]*_[0-9]*.png");
+            const arr = stdout.match(reg);
+            const filePath = path.join(path.resolve('.'), arr[0]);
+            console.log("capture file name : " + arr[0]);
```
[the remark this answers is not in the document]
Comment (YoungeunKim) [reply spec/jsSpecs/ares-device.spec.js:116]: The file name is depends on date/time, so I'd like to print the file name during TC execution.

File: spec/jsSpecs/ares-device.spec.js [hunk 31]
```diff
@@ -86,3 +90,122 @@ describe(aresCmd, function() {
...
+            const curDate = new Date();
+            expect(stdout).toContain(options.device);
+            expect(stdout).toContain(curDate.getFullYear());
+            expect(stdout).toContain("display1");
+            expect(stdout).toContain(".png");
+            expect(stdout).toContain(captureDirPath);
+            expect(fs.existsSync(captureDirPath)).toBe(true);
+
+            // eslint-disable-next-line no-useless-escape
+            const reg = new RegExp("[A-Za-z0-9]*_[A-Za-z0-9]*_[0-9]*.png");
```
[the remark this answers is not in the document]
Comment (YoungeunKim) [reply spec/jsSpecs/ares-device.spec.js:154]: I fixed it as you suggested regex.

File: spec/jsSpecs/ares-device.spec.js [hunk 32]
```diff
@@ -86,3 +90,122 @@ describe(aresCmd, function() {
...
+            expect(stdout).toContain("screen.bmp");
+            expect(stdout).toContain(captureDirPath);
+            expect(fs.existsSync(captureDirPath)).toBe(true);
+
+            done();
+        });
+    });
+
+    it('Capture with directory & file path(jpg)', function(done) {
+        const capFilePath = path.join(captureDirPath, "screen.jpg");
```
[the remark this answers is not in the document]
Comment (YoungeunKim) [reply spec/jsSpecs/ares-device.spec.js:177]: Done.

File: spec/jsSpecs/ares-device.spec.js [hunk 33]
```diff
@@ -86,3 +90,122 @@ describe(aresCmd, function() {
...
+        exec(cmd + ` --capture ${capFilePath}`, function (error, stdout) {
+            expect(stdout).not.toContain(options.device);
+            expect(stdout).not.toContain("display0");
+            expect(stdout).toContain("screen.jpg");
+            expect(stdout).toContain(captureDirPath);
+            expect(fs.existsSync(captureDirPath)).toBe(true);
+
```
[the remark this answers is not in the document]
Comment (YoungeunKim) [reply spec/jsSpecs/ares-device.spec.js:184]: Done.

File: spec/jsSpecs/ares-device.spec.js [hunk 34]
```diff
@@ -5,8 +5,12 @@
  */
 
 const exec = require('child_process').exec,
+    path = require('path'),
+    fs = require('fs'),
     common = require('./common-spec');
 
+const captureDirPath = path.join(__dirname, "..", "tempFiles", "webOSCap");
```
[the remark this answers is not in the document]
Comment (YoungeunKim) [reply spec/jsSpecs/ares-device.spec.js:12]: I replaced it to "deviceCapture"

File: lib/device.js (hunk 18 above)
Comment (YoungeunKim) [reply lib/device.js:281]: Done.

File: lib/device.js (hunk 19 above)
Comment (YoungeunKim) [reply lib/device.js:309]: Yes.

File: lib/device.js (hunk 20 above)
Comment (YoungeunKim) [reply lib/device.js:312]: Done.

File: lib/device.js (hunk 19 above)
Comment (YoungeunKim) [reply lib/device.js:309]: extname returns `.bmp`, I want to get only "bmp" without `.`

File: lib/device.js (hunk 21 above)
Comment (YoungeunKim) [reply lib/device.js:355]: It's "padding." make digit number over 2 length.

File: lib/device.js (hunk 22 above)
Comment (YoungeunKim) [reply lib/device.js:367]: _captureFile is better, cause already captureScreen () is defined in device.js

File: lib/device.js (hunk 23 above)
Comment (YoungeunKim) [reply lib/device.js:387]: Please check luna.send() in luna.js.
 The function has error handle logic when lineObj.returnValue is null or false.

File: lib/device.js (hunk 24 above)
Comment (YoungeunKim) [reply lib/device.js:411]: Done.

File: lib/device.js (hunk 25 above)
Comment (YoungeunKim) [reply lib/device.js:396]: I changed it. _copyFileToLocal

**Comment by YoungeunKim** [comment]
Code :-1: 
Verified :+1: 

I'm still working for apply code review.

File: spec/jsSpecs/ares-device.spec.js [hunk 35]
```diff
@@ -86,3 +90,122 @@ describe(aresCmd, function() {
...
+            expect(stdout).not.toContain(options.device);
+            expect(stdout).not.toContain("display0");
+            expect(stdout).toContain("screen.png");
+            expect(stdout).toContain(path.resolve('.'));
+
+            const filePath = path.join(path.resolve('.'), "screen.png");
+            expect(fs.existsSync(filePath)).toBe(true);
+
+            // remove created capture file
+            common.removeOutDir(filePath);
```
[the remark this answers is not in the document]
Comment (YoungeunKim) [reply spec/jsSpecs/ares-device.spec.js:137]: Done.

File: lib/device.js (hunk 16 above)
Comment (YoungeunKim) [reply lib/device.js:280]: I had a talk it. I changed the value to createdTmpDir.
Done.

File: lib/device.js (hunk 17 above)
Comment (YoungeunKim) [reply lib/device.js:333]: I change it to /tmp/aresCapture

File: files/help/ares-device.help (hunk 26 above)
[the remark this answers is not in the document]
Comment (YoungeunKim) [reply files/help/ares-device.help:14]: @ssuminahn , Please suggest it. If you don't give it, I'll keep current description.

[the remark this answers is not in the document]
Comment (ssuminahn) [reply files/help/ares-device.help:14]: @YoungeunKim 
There is nothing that comes to mind either. How about the sentence below?
This command displays the device information and supports screen capture.

[the remark this answers is not in the document]
Comment (YoungeunKim) [reply files/help/ares-device.help:14]: @ssuminahn , Good! I'll apply it.

**Review by YoungeunKim — COMMENTED** [review]
Code :+1: 
Verified :+1:

File: spec/jsSpecs/ares-device.spec.js [hunk 36]
```diff
@@ -93,32 +93,32 @@ describe(aresCmd, function() {
...
+    let generatedFile = "";
+
     beforeEach(function(done) {
         common.removeOutDir(captureDirPath);
         done();
     });
     afterEach(function(done) {
+        common.removeOutDir(generatedFile);
         common.removeOutDir(captureDirPath);
+        generatedFile = "";
```
Comment (ssuminahn) [inline spec/jsSpecs/ares-device.spec.js:105]: is this proper to move beforeEach?

File: spec/jsSpecs/ares-device.spec.js [hunk 37]
```diff
@@ -5,8 +5,13 @@
  */
 
 const exec = require('child_process').exec,
+    path = require('path'),
+    fs = require('fs'),
     common = require('./common-spec');
 
+const captureDirPath = path.join(__dirname, "..", "tempFiles", "deviceCapture"),
+    reg = new RegExp("[A-Za-z0-9]*_display[0-9]_[0-9]*.png");
```
Comment (ssuminahn) [inline spec/jsSpecs/ares-device.spec.js:13]: variable name

File: spec/jsSpecs/ares-device.spec.js [hunk 38]
```diff
@@ -86,3 +91,112 @@ describe(aresCmd, function() {
...
+
+    it('Capture', function(done) {
+        exec(cmd + ` -c`, function (error, stdout) {
+            expect(stdout).toContain(options.device);
+            expect(stdout).toContain(new Date().getFullYear());
+            expect(stdout).toContain("display0");
+            expect(stdout).toContain(".png");
+            expect(stdout).toContain(path.resolve('.'));
+
+            const arr = stdout.match(reg);
```
Comment (ssuminahn) [inline spec/jsSpecs/ares-device.spec.js:117]: variable name

File: spec/jsSpecs/ares-device.spec.js [hunk 39]
```diff
@@ -86,3 +91,112 @@ describe(aresCmd, function() {
...
+
+            generatedFile = path.join(path.resolve('.'), arr[0]);
+            console.log("capture file name : " + arr[0]);
```
Comment (ssuminahn) [inline spec/jsSpecs/ares-device.spec.js:120]: is this needed?

File: lib/device.js [hunk 40]
```diff
@@ -251,6 +256,168 @@ const util = require('util'),
...
+                    let parseExt = path.parse(options.outputPath).ext;
+                    parseExt = parseExt.split('.').pop();
+
+                    log.info("device#capture#_makeCaptureOption():" + " dir name : " + parseDirPath 
+                        + " ,file name : " + parseBase + " ,inputFormat : " + parseExt);
+
+                    // if given path has extension
+                    if (parseBase && parseExt) {
+                        if (parseExt === "png" || parseExt === "bmp" || parseExt === "jpg") {
+                            captureFormat = parseExt.toUpperCase();
```
Comment (ssuminahn) [inline lib/device.js:317]: why this convert to uppercase?

**Review by ssuminahn — CHANGES_REQUESTED** [review]

File: lib/device.js (hunk 40 above)
Comment (YoungeunKim) [reply lib/device.js:317]: Need to set "format" as uppercase. 

# luna-send -n 1 luna://com.webos.surfacemanager/captureCompositorOutput '{"output":"/tmp/settings.bmp","format":"bmp"}'
{
    "errorCode": 104,
    "errorText": "ERR_INVALID_FORMAT",
    "returnValue": false
}

File: lib/device.js (hunk 19 above)
Comment (YoungeunKim) [reply lib/device.js:309]: I'll keep it.

File: lib/device.js (hunk 22 above)
Comment (YoungeunKim) [reply lib/device.js:366]: captureScreenFile();

File: spec/jsSpecs/ares-device.spec.js (hunk 36 above)
Comment (YoungeunKim) [reply spec/jsSpecs/ares-device.spec.js:105]: Done.

File: spec/jsSpecs/ares-device.spec.js (hunk 37 above)
Comment (YoungeunKim) [reply spec/jsSpecs/ares-device.spec.js:13]: Done.

File: spec/jsSpecs/ares-device.spec.js (hunk 38 above)
Comment (YoungeunKim) [reply spec/jsSpecs/ares-device.spec.js:117]: Done.

File: spec/jsSpecs/ares-device.spec.js (hunk 39 above)
Comment (YoungeunKim) [reply spec/jsSpecs/ares-device.spec.js:120]: Yes.

File: lib/device.js (hunk 22 above)
Comment (YoungeunKim) [reply lib/device.js:366]: Done.

**Comment by YoungeunKim** [comment]
Code :+1: 
Verified :+1:

**Comment by ssuminahn** [comment]
code :+1: 
verify :+1:

**Review by ssuminahn — COMMENTED** [review]
code :+1: 
verify :+1:

**Review by ssuminahn — COMMENTED** [review]
code :+1: 
verify :+1:

**Review by ssuminahn — APPROVED** [review]

**Pull request merged**